jetson/percep/main.cpp

```diff
@@ -10,176 +10,177 @@ using namespace std::chrono_literals;
+        #if AR_DETECTION
+            tagPair = detector.findARTags(src, depth_img, rgb);
+            #if AR_RECORD
+                cam.record_ar(rgb);
+            #endif
+
+            detector.updateDetectedTagInfo(arTags, tagPair, depth_img, src);
+
+        #if PERCEPTION_DEBUG && AR_DETECTION
+            imshow("depth", src);
+            waitKey(1);  
+        #endif
+
+        #endif
+
+        /* --- Point Cloud Processing --- */
+        #if OBSTACLE_DETECTION && !WRITE_CURR_FRAME_TO_DISK
+        
+        #if PERCEPTION_DEBUG
+            //Update Original 3D Viewer
+            pointcloud.updateViewer(originalView);
+            cout<<"Original W: " <<pointcloud.pt_cloud_ptr->width<<" Original H: "<<pointcloud.pt_cloud_ptr->height<<endl;
+        #endif
+
+        //Run Obstacle Detection
+        pointcloud.pcl_obstacle_detection();  
+        obstacle_return obstacle_detection (pointcloud.bearing, pointcloud.distance);
+
+        //Outlier Detection Processing
+        outliers.pop_back(); //Remove outdated outlier value
+
+        if(pointcloud.bearing > 0.05 || pointcloud.bearing < -0.05)
+            outliers.push_front(true);//if an obstacle is detected in front
+        else 
+            outliers.push_front(false); //obstacle is not detected
+
+        if(outliers == checkTrue) //If past iterations see obstacles
+            lastObstacle = obstacle_detection;
+        else if (outliers == checkFalse) // If our iterations see no obstacles after seeing obstacles
+            lastObstacle = obstacle_detection;
+
+        obstacleMessage.distance = lastObstacle.distance; //update LCM distance field
+        obstacleMessage.bearing = lastObstacle.bearing; //update LCM bearing field
+        #if PERCEPTION_DEBUG
+            cout << "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!Path Sent: " << obstacleMessage.bearing << "\n";
+            cout << "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!Distance Sent: " << obstacleMessage.distance << "\n";
+        #endif
+
+        #if PERCEPTION_DEBUG
+        //Update Processed 3D Viewer
+        pointcloud.updateViewer(newView);
+        #if PERCEPTION_DEBUG
+            cout<<"Downsampled W: " <<pointcloud.pt_cloud_ptr->width<<" Downsampled H: "<<pointcloud.pt_cloud_ptr->height<<endl;
+        #endif
+        #endif
+        
+        #endif
+        
+        /* --- Publish LCMs --- */
+        lcm_.publish("/target_list", &arTagsMessage);
+        lcm_.publish("/obstacle", &obstacleMessage);
+
+        #if !ZED_SDK_PRESENT
+            std::this_thread::sleep_for(0.2s); // Iteration speed control not needed when using camera 
+        #endif
+        
+        ++iterations;
   }
 
 
-/* --- Wrap Things Up --- */
-  #if OBSTACLE_DETECTION && PERCEPTION_DEBUG
-    viewer->close();
-  #endif
+    /* --- Wrap Things Up --- */
+    #if OBSTACLE_DETECTION && PERCEPTION_DEBUG
+        pointcloud.~PCL();
```

Comment:
I believe the destructor is going to be called once the PCL object falls out of scope. I don't think that this #if is actually doing anything. If you want the destructor to do nothing when PERCPTION_DEBUG=true you should add the directives in the PCL destructor itsefl

Reply:
Fixed